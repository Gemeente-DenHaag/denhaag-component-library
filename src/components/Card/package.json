{
  "name": "@gemeente-denhaag/card",
  "version": "0.1.7",
  "description": "A Card component",
  "bugs": "https://github.com/Gemeente-DenHaag/denhaag-component-library/issues",
  "repository": {
    "type": "git",
    "url": "https://github.com/Gemeente-DenHaag/denhaag-component-library.git",
    "directory": "src/components/Card"
  },
  "license": "EUPL-1.2",
  "author": "Municipality of The Hague",
  "exports": "./dist/index.js",
  "main": "dist/index.js",
  "module": "dist/index.js",
  "types": "dist/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "rollup -c ../../../rollup.config.js",
    "clean": "yarn rimraf dist tsconfig.tsbuildinfo"
  },
  "dependencies": {
    "@gemeente-denhaag/baseprops": "^0.1.7",
    "@material-ui/core": "^4.11.0",
    "@gemeente-denhaag/icons": "^0.1.0"
  },
  "peerDependencies": {
    "react": "^17.0.1"
  },
  "devDependencies": {
    "@gemeente-denhaag/avatar": "*",
    "@gemeente-denhaag/iconbutton": "*"
  },
  "gitHead": "dcf72a9b79266c1ebede35aff4a02dd9121a980f"
}
