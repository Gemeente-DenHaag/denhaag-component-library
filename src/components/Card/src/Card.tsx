import React from "react";
import MaterialCard from "@material-ui/core/Card";
import { BasePaperProps } from "./BasePaperProps";

export interface CardProps extends BasePaperProps {
  /**
   * If `true`, the card will use raised styling.
   */
  raised?: boolean;
}

/**
 * Primary UI component for user interaction
 */
export const Card: React.FC<CardProps> = (props: CardProps) => {
  return <MaterialCard {...props}>{props.children}</MaterialCard>;
};

/**
 * Default export for Card
 */
export default Card;
