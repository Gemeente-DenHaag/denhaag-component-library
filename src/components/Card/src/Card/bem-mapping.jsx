export const cardClasses = {
  root: 'denhaag-card',
};

export const cardArrowClasses = {
  root: 'denhaag-card__arrow',
};

export const cardTitleClasses = {
  root: 'denhaag-card__title',
};

export const cardSubtitleClasses = {
  root: 'denhaag-card__subtitle',
};

export const cardPaperClasses = {
  root: 'denhaag-card__paper',
};

export const cardCaseClasses = {
  root: 'denhaag-card denhaag-card--case',
};
