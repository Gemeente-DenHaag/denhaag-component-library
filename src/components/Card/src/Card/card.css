.denhaag-card {
  border-radius: var(--denhaag-card-border-radius);
  width: var(--denhaag-card-width);
}

.denhaag-card--focus,
.denhaag-card:focus-within {
  outline: none;
}

.denhaag-card--hover,
.denhaag-card:hover,
.denhaag-card:hover:focus {
  cursor: pointer;
}

.denhaag-card__wrapper {
  border-color: var(--denhaag-card-border-color);
  border-radius: var(--denhaag-card-border-radius);
  border-style: solid;
  border-width: var(--denhaag-card-border-width);
  box-sizing: border-box;
  display: flex;
  flex-direction: column;
  height: var(--denhaag-card-height);
  padding-block-start: var(--denhaag-card-wrapper-padding-block);
  padding-block-end: var(--denhaag-card-wrapper-padding-block);
  padding-inline-start: var(--denhaag-card-wrapper-padding-inline);
  padding-inline-end: var(--denhaag-card-wrapper-padding-inline);
}

.denhaag-card--focus .denhaag-card__text-wrapper,
.denhaag-card:focus-within .denhaag-card__text-wrapper {
  outline: var(--denhaag-focus-border);
  position: relative;
}

.denhaag-card__arrow {
  left: var(--denhaag-card-arrow-left);
  position: var(--denhaag-card-arrow-position);
  right: var(--denhaag-card-arrow-right);
}

.denhaag-card__title {
  color: var(--denhaag-title-font-color);
  font-size: var(--denhaag-card-title-font-size);
  line-height: var(--denhaag-card-title-line-height);
  padding-block-end: var(--denhaag-card-title-padding-block);
  padding-block-start: var(--denhaag-card-title-padding-block);
  padding-inline-end: var(--denhaag-card-title-padding-block);
  padding-inline-start: var(--denhaag-card-title-padding-inline);
  text-decoration: none;
}

.denhaag-card__title a:link,
.denhaag-card__title a:visited {
  color: inherit;
  text-decoration: none;
}

.denhaag-card__title a:link:focus {
  outline: none;
}

.denhaag-card__subtitle {
  color: var(--denhaag-card-subtitle-font-color);
  font-size: var(--denhaag-card-subtitle-font-size);
  font-weight: var(--denhaag-card-subtitle-font-weight);
  line-height: var(--denhaag-card-subtitle-line-height);
  padding-block-end: var(--denhaag-card-subtitle-padding-block);
  padding-block-start: var(--denhaag-card-subtitle-padding-block);
  padding-inline-end: var(--denhaag-card-subtitle-padding-block);
  padding-inline-start: var(--denhaag-card-subtitle-padding-block);
}

.denhaag-card--case {
  background: var(--denhaag-card-case-secondary-background-color);
  padding-block-start: var(--denhaag-card-case-padding-block-start);
  position: relative;
}

.denhaag-card--case::before {
  background: var(--denhaag-card-case-paper-color);
  border-radius: 0 var(--denhaag-card-border-radius) 0 0;
  content: "";
  display: block;
  height: 20px;
  position: absolute;
  right: 8px;
  top: 8px;
  width: 50%;
}

.denhaag-card--case .denhaag-card__wrapper {
  background: transparent;
  border-radius: var(--denhaag-card-border-radius);
  border: 2px;
  box-sizing: border-box;
  display: flex;
  flex-direction: column;
  height: var(--denhaag-card-case-height);
  padding-block-end: var(--denhaag-card-case-padding-block-end);
  padding-block-start: var(--denhaag-card-case-padding-block-start);
  perspective: 1800px;
  position: relative;
}

.denhaag-card--case .denhaag-card__background {
  background: var(--denhaag-card-case-primary-background-color);
  border-radius: var(--denhaag-card-border-radius);
  border-top-left-radius: 0;
  bottom: 0;
  left: 0;
  position: absolute;
  right: 0;
  top: 0;
  transform-origin: 50% 100%;
  z-index: -1;
}

.denhaag-card--case .denhaag-card__background::before {
  background: var(--denhaag-card-case-primary-background-color);
  border-top-left-radius: var(--denhaag-card-border-radius);
  content: "";
  display: block;
  height: 25px;
  left: 0;
  position: absolute;
  top: -20px;
  width: 60%;
}

.denhaag-card--case .denhaag-card__background::after {
  background: var(--denhaag-card-case-primary-background-color);
  content: "";
  display: block;
  height: 22px;
  left: 55%;
  position: absolute;
  top: -20px;
  transform: skewX(30deg);
  width: 10%;
}

.denhaag-card--case .denhaag-card__arrow {
  color: var(--denhaag-card-case-color);
  left: var(--denhaag-card-arrow-left);
  position: var(--denhaag-card-arrow-position);
  right: var(--denhaag-card-arrow-right);
}

.denhaag-card--case--hover .denhaag-card__arrow,
.denhaag-card--case--focus .denhaag-card__arrow,
.denhaag-card--case:hover .denhaag-card__arrow,
.denhaag-card--case:focus-within .denhaag-card__arrow {
  color: var(--denhaag-card-case-active-color);
}

/* stylelint-disable-next-line no-descending-specificity */
.denhaag-card--focus .denhaag-card__arrow,
.denhaag-card:focus-within .denhaag-card__arrow {
  outline: var(--denhaag-focus-border);
}

.denhaag-card--case .denhaag-card__title {
  position: relative;
  color: var(--denhaag-card-case-color);
  font-weight: 700;
  padding-block-end: var(--denhaag-card-title-padding-block);
  padding-block-start: var(--denhaag-card-title-padding-block);
  padding-inline-end: var(--denhaag-card-title-padding-inline);
  padding-inline-start: var(--denhaag-card-title-padding-inline);
}

.denhaag-card--case--hover .denhaag-card__title,
.denhaag-card--case--focus .denhaag-card__title,
.denhaag-card--case:hover .denhaag-card__title,
.denhaag-card--case:focus-within .denhaag-card__title {
  color: var(--denhaag-card-case-active-color);
}

.denhaag-card--case .denhaag-card__subtitle {
  color: var(--denhaag-card-case-color);
  font-weight: 500;
  padding-block-start: var(--denhaag-card-subtitle-padding);
  padding-block-end: var(--denhaag-card-subtitle-padding);
  padding-inline-start: var(--denhaag-card-subtitle-padding);
  padding-inline-end: var(--denhaag-card-subtitle-padding);
  position: relative;
}

.denhaag-card--case--hover .denhaag-card__subtitle,
.denhaag-card--case--focus .denhaag-card__subtitle,
.denhaag-card--case:hover .denhaag-card__subtitle,
.denhaag-card--case:focus-within .denhaag-card__subtitle {
  color: var(--denhaag-card-case-active-color);
}

.denhaag-card--case--hover::before,
.denhaag-card--case--focus::before,
.denhaag-card--case:hover::before,
.denhaag-card--case:focus-within::before {
  top: 12px;
  transform-origin: right;
  transform: rotate(-2.45deg);
}

.denhaag-card--case--focus .denhaag-card__background,
.denhaag-card--case--hover .denhaag-card__background,
.denhaag-card--case:hover .denhaag-card__background,
.denhaag-card--case:focus-within .denhaag-card__background {
  transform: rotateX(-20deg);
}
