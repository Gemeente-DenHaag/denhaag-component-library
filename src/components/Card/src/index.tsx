import React from "react";
import MaterialCard from "@material-ui/core/Card";
import { BasePaperProps } from "@gemeente-denhaag/paper";

export interface CardProps extends BasePaperProps {
  /**
   * If `true`, the card will use raised styling.
   */
  raised?: boolean;
}

/**
 * Primary UI component for user interaction
 */
const Card: React.FC<CardProps> = (props: CardProps) => {
  return <MaterialCard {...props}>{props.children}</MaterialCard>;
};

/**
 * Default export for Card
 */
export default Card;
export * from "@gemeente-denhaag/cardactions";
export * from "@gemeente-denhaag/cardcontent";
export * from "@gemeente-denhaag/cardheader";
export * from "@gemeente-denhaag/paper";
