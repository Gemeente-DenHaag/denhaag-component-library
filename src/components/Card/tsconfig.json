{
  "extends": "../../../tsconfig.build.json",
  "compilerOptions": {
    "rootDir": "src",
    "outDir": "dist",
    "composite": true
  },
  "include": ["src/**/*"],
  "exclude": ["dist", "src/**/*.stories.tsx"],
  "references": [
    {
      "path": "../Avatar"
    },
    {
      "path": "../IconButton"
    },
    {
      "path": "../Icons"
    },
    {
      "path": "../BaseProps"
    }
  ]
}
