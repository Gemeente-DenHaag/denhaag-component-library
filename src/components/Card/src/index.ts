import Card from './Card/Card';

export default Card;
export * from './Card/Card';

export * from './Paper/Paper';
