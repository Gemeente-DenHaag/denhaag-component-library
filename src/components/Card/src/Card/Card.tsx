import React, { createRef } from 'react';
import { Card as MaterialCard, Icon, Typography } from '@material-ui/core';
import { ArrowRightIcon } from '@gemeente-denhaag/icons';
import BaseProps from '@gemeente-denhaag/baseprops';
import './mui-override.css';
import './card.css';
import { cardArrowClasses, cardCaseClasses, cardClasses, cardSubtitleClasses, cardTitleClasses } from './bem-mapping';
import { CardContent } from '../CardContent/CardContent';
import { CardActions } from '../CardActions/CardActions';
import clsx from 'clsx';

export interface CardProps extends Omit<BaseProps, 'classes'> {
  /**
   * Callback fired when the Card is clicked.
   */
  onClick?: (event: React.MouseEvent<HTMLDivElement>) => void;

  /**
   * Determines the variant of the card
   * Variant can be either `basic` or `case`
   */
  variant?: 'basic' | 'case';

  /**
   * Determines the title of the card
   */
  title: string;

  /**
   * Determines the subtitle of the card
   */
  subTitle?: string;

  /**
   * Determines the date shown on the card
   */
  date: Date;

  /**
   * Determines the url the card points to
   */
  href?: string;

  /**
   * Set to false to indicate an archived or inactive item
   */
  active?: boolean;
}

/**
 * Primary UI component for user interaction
 */
export const Card: React.FC<CardProps> = ({ active = true, ...props }: CardProps) => {
  const rootClassNames = clsx(
    cardClasses.root,
    props.variant === 'case' && cardCaseClasses.root,
    active === false && 'denhaag-card--inactive',
    props.className,
  );
  const arrowClasses = cardArrowClasses;
  const titleClasses = cardTitleClasses;
  const subtitleClasses = cardSubtitleClasses;
  const linkRef = createRef<HTMLAnchorElement>();

  const title = (
    <a href={props.href} ref={linkRef}>
      {props.title}
    </a>
  );

  const onClick = (event: React.MouseEvent<HTMLDivElement>) => {
    if (typeof props.onClick === 'function') {
      props.onClick(event);
      event.stopPropagation();
    } else if (linkRef.current !== null) {
      linkRef.current.click();
    }
  };

  return (
    <MaterialCard className={rootClassNames} onClick={onClick}>
      <div className="denhaag-card__wrapper">
        <div className="denhaag-card__background"></div>
        <CardContent>
          <div className="denhaag-card__text-wrapper">
            <Typography classes={titleClasses} component="p">
              {title}
            </Typography>
            <Typography classes={subtitleClasses} component="p">
              {props.subTitle}
            </Typography>
          </div>
          <CardActions disableSpacing={true}>
            <Typography component="div" classes={subtitleClasses}>
              <time dateTime={props.date.toDateString()}>{props.date.toLocaleDateString()}</time>
            </Typography>
            <Icon classes={arrowClasses} aria-label="ArrowRightIcon">
              <ArrowRightIcon />
            </Icon>
          </CardActions>
        </CardContent>
      </div>
    </MaterialCard>
  );
};

/**
 * Default export for Card
 */
export default Card;
