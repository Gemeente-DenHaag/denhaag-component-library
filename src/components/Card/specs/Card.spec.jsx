import * as React from 'react';
import Card from '../src/index';

const props = {
  title: 'Shrimp and Chorizo Paella',
  subTitle: 'This impressive paella is a perfect party dish and a fun meal to cook.',
  date: new Date('2020-01-21'),
  href: '#test',
};

describe(`Card tests`, () => {
  it('can be clicked', () => {
    cy.mount(<Card {...props} />);

    cy.get('.denhaag-card').click();

    cy.location('hash').should('eq', '#test');
  });

  it('can be operated with keyboard', () => {
    cy.mount(<Card {...props} />);

    cy.get('.denhaag-card a').type('{enter}');

    cy.location('hash').should('eq', '#test');
  });

  it('a11y and snapshots', () => {
    cy.snapshots(
      <Card {...props} />,
      { variant: ['case'] },
      {
        selector: '.denhaag-card',
        states: {
          hover: 'denhaag-card--hover',
          focus: 'denhaag-card--focus',
        },
      },
      [{ id: 'color-contrast', issue: 376 }],
    );
  });
});
