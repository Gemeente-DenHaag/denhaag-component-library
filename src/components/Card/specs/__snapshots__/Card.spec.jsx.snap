exports[`Card tests > a11y and snapshots #0`] = `
<div id="wrapper" style="display: inline-block;">
  <div style="padding: 10px;">
    <div
      class="MuiPaper-root MuiCard-root denhaag-card MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div class="denhaag-card__wrapper">
        <div class="denhaag-card__background"></div>
        <div class="MuiCardContent-root denhaag-card__content">
          <div class="denhaag-card__text-wrapper">
            <p
              class="MuiTypography-root denhaag-card__title MuiTypography-body1"
            >
              <a href="#test">Shrimp and Chorizo Paella</a>
            </p>
            <p
              class="MuiTypography-root denhaag-card__subtitle MuiTypography-body1"
            >
              This impressive paella is a perfect party dish and a fun meal to
              cook.
            </p>
          </div>
          <div class="MuiCardActions-root denhaag-card__actions">
            <div
              class="MuiTypography-root denhaag-card__subtitle MuiTypography-body1"
            >
              <time datetime="2020-01-21T00:00:00.000Z">1/21/2020</time>
            </div>
            <span
              class="material-icons MuiIcon-root denhaag-card__arrow"
              aria-hidden="true"
              aria-label="ArrowRightIcon"
              ><svg
                width="1em"
                height="1em"
                viewBox="0 0 24 24"
                fill="none"
                xmlns="http://www.w3.org/2000/svg"
                class="MuiSvgIcon-root denhaag-icon"
                focusable="false"
                aria-hidden="true"
                shape-rendering="auto"
              >
                <path
                  d="M12.293 5.293a1 1 0 011.414 0l6 6a1 1 0 010 1.414l-6 6a1 1 0 01-1.414-1.414L16.586 13H5a1 1 0 110-2h11.586l-4.293-4.293a1 1 0 010-1.414z"
                  fill="currentColor"
                ></path></svg
            ></span>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div style="padding: 10px;">
    <div
      class="MuiPaper-root MuiCard-root denhaag-card denhaag-card--case MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div class="denhaag-card__wrapper">
        <div class="denhaag-card__background"></div>
        <div class="MuiCardContent-root denhaag-card__content">
          <div class="denhaag-card__text-wrapper">
            <p
              class="MuiTypography-root denhaag-card__title MuiTypography-body1"
            >
              <a href="#test">Shrimp and Chorizo Paella</a>
            </p>
            <p
              class="MuiTypography-root denhaag-card__subtitle MuiTypography-body1"
            >
              This impressive paella is a perfect party dish and a fun meal to
              cook.
            </p>
          </div>
          <div class="MuiCardActions-root denhaag-card__actions">
            <div
              class="MuiTypography-root denhaag-card__subtitle MuiTypography-body1"
            >
              <time datetime="2020-01-21T00:00:00.000Z">1/21/2020</time>
            </div>
            <span
              class="material-icons MuiIcon-root denhaag-card__arrow"
              aria-hidden="true"
              aria-label="ArrowRightIcon"
              ><svg
                width="1em"
                height="1em"
                viewBox="0 0 24 24"
                fill="none"
                xmlns="http://www.w3.org/2000/svg"
                class="MuiSvgIcon-root denhaag-icon"
                focusable="false"
                aria-hidden="true"
                shape-rendering="auto"
              >
                <path
                  d="M12.293 5.293a1 1 0 011.414 0l6 6a1 1 0 010 1.414l-6 6a1 1 0 01-1.414-1.414L16.586 13H5a1 1 0 110-2h11.586l-4.293-4.293a1 1 0 010-1.414z"
                  fill="currentColor"
                ></path></svg
            ></span>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div style="padding: 10px;">
    <div
      class="MuiPaper-root MuiCard-root denhaag-card denhaag-card--case denhaag-card--archived MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div class="denhaag-card__wrapper">
        <div class="denhaag-card__background"></div>
        <div class="MuiCardContent-root denhaag-card__content">
          <div class="denhaag-card__text-wrapper">
            <p
              class="MuiTypography-root denhaag-card__title MuiTypography-body1"
            >
              <a href="#test">Shrimp and Chorizo Paella</a>
            </p>
            <p
              class="MuiTypography-root denhaag-card__subtitle MuiTypography-body1"
            >
              This impressive paella is a perfect party dish and a fun meal to
              cook.
            </p>
          </div>
          <div class="MuiCardActions-root denhaag-card__actions">
            <div
              class="MuiTypography-root denhaag-card__subtitle MuiTypography-body1"
            >
              <time datetime="2020-01-21T00:00:00.000Z">1/21/2020</time>
            </div>
            <span
              class="material-icons MuiIcon-root denhaag-card__arrow"
              aria-hidden="true"
              aria-label="ArrowRightIcon"
              ><svg
                width="1em"
                height="1em"
                viewBox="0 0 24 24"
                fill="none"
                xmlns="http://www.w3.org/2000/svg"
                class="MuiSvgIcon-root denhaag-icon"
                focusable="false"
                aria-hidden="true"
                shape-rendering="auto"
              >
                <path
                  d="M12.293 5.293a1 1 0 011.414 0l6 6a1 1 0 010 1.414l-6 6a1 1 0 01-1.414-1.414L16.586 13H5a1 1 0 110-2h11.586l-4.293-4.293a1 1 0 010-1.414z"
                  fill="currentColor"
                ></path></svg
            ></span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
